_pytest/mark.py

```diff
@@ -329,30 +336,40 @@ def __call__(self, *args, **kwargs):
             is_class = inspect.isclass(func)
             if len(args) == 1 and (istestfunc(func) or is_class):
                 if is_class:
-                    if hasattr(func, 'pytestmark'):
-                        mark_list = func.pytestmark
-                        if not isinstance(mark_list, list):
-                            mark_list = [mark_list]
-                        # always work on a copy to avoid updating pytestmark
-                        # from a superclass by accident
-                        mark_list = mark_list + [self]
-                        func.pytestmark = mark_list
-                    else:
-                        func.pytestmark = [self]
+                    apply_mark(func, self.mark)
                 else:
-                    holder = getattr(func, self.name, None)
-                    if holder is None:
-                        holder = MarkInfo(self.mark)
-                        setattr(func, self.name, holder)
-                    else:
-                        holder.add_mark(self.mark)
+                    apply_legacy_mark(func, self.mark)
+                    apply_mark(func, self.mark)
                 return func
 
         mark = Mark(self.name, args, kwargs)
         return self.__class__(self.mark.combined_with(mark))
 
 
+def apply_mark(obj, mark):
+    assert isinstance(mark, Mark), mark
+    """applies a marker to an object,
```

Comment:
> im against a get_pristine_marks function, - the marks api people use to get/interact with marks sould be an new object on item that we will provide

I was actually suggesting using `get_pristine_marks` *instead* of `pytestmark`. Certainly accessing the merged marks will done through `Item`'s API. But after I concluded later it probably won't make much sense given that for modules at least the `pytestmark` mechanism to *declare* marks will have to stay, so we might as well use the same for classes and functions as well. So let's move this forward as it is. 😁 

**Side topic**: usually when things decorate user code, it might be a better to use a public API to manipulate what's being decorated instead of making the decorated attribute public. For example the `mock` package experienced user experience problems because you can add arbitrary attributes to `Mock` objects *and* use the same objects to use functionality, for example `Mock().assert_called_once()`. The problem with that is that it was common for users to mistype some name and it would silently do nothing (it would return a new `Mock` instance) as opposed to a failing error. This could have been better addressed I think by providing that functionality through an API instead through the object directly, for example `mock.assert_called_once(m)` (with `mock` here being the module and `m` the `Mock` object), which would prevent clashes and mistakes. 

That was the general feeling that I had when proposing to use an API to access the list of unmerged marks: it would allow us to use an internal ugly name, possibly change this in the future to store that information somewhere without breaking user expectations, and even do some more work instead of just returning an internal attribute of the object.

But as I said, let's move forward with `pytestmark` then. 👍